xwayland/xwm.c

```diff
@@ -492,84 +614,282 @@ static void handle_configure_request(struct wlr_xwm *xwm,
 	wlr_log(L_DEBUG, "XCB_PROPERTY_NOTIFY (%u)", ev->window);
-	struct wlr_xwayland_surface *surface = lookup_surface_any(xwm, ev->window);
+	struct wlr_xwayland_surface *surface = lookup_surface(xwm, ev->window);
 	if (surface == NULL) {
 		return;
 	}
 
 	read_surface_property(xwm, surface, ev->atom);
 }
 
+static void handle_surface_id_message(struct wlr_xwm *xwm,
+		xcb_client_message_event_t *ev) {
+	struct wlr_xwayland_surface *xsurface = lookup_surface(xwm, ev->window);
+	if (xsurface == NULL) {
+		wlr_log(L_DEBUG,
+			"client message WL_SURFACE_ID but no new window %u ?",
+			ev->window);
+		return;
+	}
+	/* Check if we got notified after wayland surface create event */
+	uint32_t id = ev->data.data32[0];
+	struct wl_resource *resource =
+		wl_client_get_object(xwm->xwayland->client, id);
+	if (resource) {
+		struct wlr_surface *surface = wl_resource_get_user_data(resource);
+		xsurface->surface_id = 0;
+		map_shell_surface(xwm, xsurface, surface);
+	} else {
+		xsurface->surface_id = id;
+		wl_list_insert(&xwm->unpaired_surfaces, &xsurface->unpaired_link);
+	}
+}
+
+#define _NET_WM_MOVERESIZE_SIZE_TOPLEFT      0
+#define _NET_WM_MOVERESIZE_SIZE_TOP          1
+#define _NET_WM_MOVERESIZE_SIZE_TOPRIGHT     2
+#define _NET_WM_MOVERESIZE_SIZE_RIGHT        3
+#define _NET_WM_MOVERESIZE_SIZE_BOTTOMRIGHT  4
+#define _NET_WM_MOVERESIZE_SIZE_BOTTOM       5
+#define _NET_WM_MOVERESIZE_SIZE_BOTTOMLEFT   6
+#define _NET_WM_MOVERESIZE_SIZE_LEFT         7
+#define _NET_WM_MOVERESIZE_MOVE              8  // movement only
+#define _NET_WM_MOVERESIZE_SIZE_KEYBOARD     9  // size via keyboard
+#define _NET_WM_MOVERESIZE_MOVE_KEYBOARD    10  // move via keyboard
+#define _NET_WM_MOVERESIZE_CANCEL           11  // cancel operation
+
+static void handle_net_wm_moveresize_message(struct wlr_xwm *xwm,
+		xcb_client_message_event_t *ev) {
+	// same as xdg-toplevel-v6
+	// TODO need a common enum for this
+	static const int map[] = {
+		5, 1, 9, 8, 10, 2, 6, 4
+	};
+
+	struct wlr_xwayland_surface *xsurface = lookup_surface(xwm, ev->window);
+	if (!xsurface) {
+		return;
+	}
+
+	// TODO: we should probably add input or seat info to this but we would just
+	// be guessing
+	struct wlr_xwayland_resize_event resize_event;
+	struct wlr_xwayland_move_event move_event;
+
+	int detail = ev->data.data32[2];
+	switch (detail) {
+	case _NET_WM_MOVERESIZE_MOVE:
+		move_event.surface = xsurface;
+		wl_signal_emit(&xsurface->events.request_move, &move_event);
+		break;
+	case _NET_WM_MOVERESIZE_SIZE_TOPLEFT:
+	case _NET_WM_MOVERESIZE_SIZE_TOP:
+	case _NET_WM_MOVERESIZE_SIZE_TOPRIGHT:
+	case _NET_WM_MOVERESIZE_SIZE_RIGHT:
+	case _NET_WM_MOVERESIZE_SIZE_BOTTOMRIGHT:
+	case _NET_WM_MOVERESIZE_SIZE_BOTTOM:
+	case _NET_WM_MOVERESIZE_SIZE_BOTTOMLEFT:
+	case _NET_WM_MOVERESIZE_SIZE_LEFT:
+		resize_event.surface = xsurface;
+		resize_event.edges = map[detail];
+		wl_signal_emit(&xsurface->events.request_resize, &resize_event);
+		break;
+	case _NET_WM_MOVERESIZE_CANCEL:
+		break;
```

Comment:
Is this a TODO?

Reply:
I don't want to handle it without a specific example of something that depends on it to function correctly.

Reply:
Can you add a comment to that effect?